### ISSUE TITLE
mro skips over first element of linearized isa array

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11481]
Migrated from [rt.perl.org#94306](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=94306) (status was 'resolved')

Searchable as RT94306$

**Comment by p5pRT** [comment]
### From @salva

#### Created by @salva



The current mro machinery unexpectedly skips over the first element of the array returned by the "resolve" callback during method resolution\.


Sample program&#8203;:
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\#\!/usr/bin/perl

use 5\.010001;
use strict;
use warnings;
use mro;
use MRO&#8203;::Define;

no strict 'refs';
MRO&#8203;::Define&#8203;::register\_mro\(mro1 => sub \{ \[ undef\, @&#8203;\{$\_\[0\]\->\{ISA\}\} \] \}\);
MRO&#8203;::Define&#8203;::register\_mro\(mro2 => sub \{ \[ @&#8203;\{$\_\[0\]\->\{ISA\}\} \] \}\);

package A;
sub foo \{ say "foo" \};

package B1;
our @&#8203;ISA = 'A';
mro&#8203;::set\_mro\(B1 => 'mro1'\);

package B2;
our @&#8203;ISA = 'A';
mro&#8203;::set\_mro\(B2 => 'mro2'\);

B1\->foo;
B2\->foo; \# unexpectedly skips A&#8203;::foo

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

I understand that the first element is skipped on purpose as perl looks for the method in the object stash before going the mro route\, but IMO that behavior should be documented \(by a native English speaker\, not myself :\-\)


A mro implementer may like to not include the object stash as the first element of the linearized @&#8203;ISA if for instance he knows it is an empty one\.\.\.


That's exactly what I was doing in a real mro implementation because I was linearizing @&#8203;ISA in the following fashion&#8203;:


  sub my\_mro \{ \[grep mro&#8203;::get\_pkg\_gen\($\_\)\, \_linearized\_isa\(@&#8203;\_\)\] \}
and I have had a hard time finding the root of the obscure bug that caused some methods to mysteriously disappear\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.12.4:

Configured by Debian Project at Thu Jul  7 09:07:11 UTC 2011.

Summary of my perl5 (revision 5 version 12 subversion 4) configuration:
   
  Platform:
    osname=linux, osvers=2.6.24-27-server, archname=x86_64-linux-gnu-thread-multi
    uname='linux yellow 2.6.24-27-server #1 smp fri mar 12 01:23:09 utc 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.12 -Darchlib=/usr/lib/perl/5.12 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.12.4 -Dsitearch=/usr/local/lib/perl/5.12.4 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.12.4 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib/x86_64-linux-gnu /lib64 /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.12.4
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.4:
    /etc/perl
    /usr/local/lib/perl/5.12.4
    /usr/local/share/perl/5.12.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.12
    /usr/share/perl/5.12
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.12.3
    /usr/local/share/perl/5.12.3
    .


Environment for perl 5.12.4:
    HOME=/home/salva
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jul 07 10&#8203;:31&#8203;:47 2011\, salva wrote&#8203;:
> The current mro machinery unexpectedly skips over the first element of
>    the array returned by the "resolve" callback during method
>    resolution\.
> 
> 
> Sample program&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \#\!/usr/bin/perl
> 
> use 5\.010001;
> use strict;
> use warnings;
> use mro;
> use MRO&#8203;::Define;
> 
> no strict 'refs';
> MRO&#8203;::Define&#8203;::register\_mro\(mro1 => sub \{ \[ undef\, @&#8203;\{$\_\[0\]\->\{ISA\}\} \] \}\);
> MRO&#8203;::Define&#8203;::register\_mro\(mro2 => sub \{ \[ @&#8203;\{$\_\[0\]\->\{ISA\}\} \] \}\);
> 
> package A;
> sub foo \{ say "foo" \};
> 
> package B1;
> our @&#8203;ISA = 'A';
> mro&#8203;::set\_mro\(B1 => 'mro1'\);
> 
> package B2;
> our @&#8203;ISA = 'A';
> mro&#8203;::set\_mro\(B2 => 'mro2'\);
> 
> B1\->foo;
> B2\->foo; \# unexpectedly skips A&#8203;::foo
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> I understand that the first element is skipped on purpose as perl
>    looks for the method in the object stash before going the mro
>    route\, but IMO that behavior should be documented \(by a native
>    English speaker\, not myself :\-\)
> 
> 
> A mro implementer may like to not include the object stash as the
>    first element of the linearized @&#8203;ISA if for instance he knows it is
>    an empty one\.\.\.
> 
> 
> That's exactly what I was doing in a real mro implementation because I
>    was linearizing @&#8203;ISA in the following fashion&#8203;:
> 
> 
> ï¿½ sub my\_mro \{ \[grep mro&#8203;::get\_pkg\_gen\($\_\)\, \_linearized\_isa\(@&#8203;\_\)\] \}
> and I have had a hard time finding the root of the obscure bug that
>    caused some methods to mysteriously disappear\.

Iâm wondering whether Perl should check whether the first element is the
same name before skipping it\. Or maybe Perl should never skip the first
name\. I need to go back and re\-read that code some time\. \(I knew it
inside\-out six months ago\.\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rafl

"Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> Iâm wondering whether Perl should check whether the first element is
> the same name before skipping it\. Or maybe Perl should never skip the
> first name\.

Either of those things would make implementing modules such as
MRO&#8203;::Magic a whole lot easier\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jul 10 13&#8203;:18&#8203;:54 2011\, flora wrote&#8203;:
> "Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > Iâm wondering whether Perl should check whether the first element is
> > the same name before skipping it\. Or maybe Perl should never skip the
> > first name\.
> 
> Either of those things would make implementing modules such as
> MRO&#8203;::Magic a whole lot easier\.

Checking whether the first element is the same as the stash name should
be done in mro\_get\_linear\_isa\, as there are various parts of the mro
code that rely on it\.  If it is not the same\, the array should be
copied\, not modified\, as the mro plugin may be using it for its own
purposes\.  The copied array can have the stash name inserted at the
beginning\.

Iâm holding off on doing this until after support for Unicode package
names is merged\.

So Iâm leaving this note here\, in case I forget and someone else wants
to tackle this\.

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed with commit 02cab67\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jul 10 13&#8203;:18&#8203;:54 2011\, flora wrote&#8203;:
> "Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > Iâm wondering whether Perl should check whether the first element is
> > the same name before skipping it\. Or maybe Perl should never skip the
> > first name\.
> 
> Either of those things would make implementing modules such as
> MRO&#8203;::Magic a whole lot easier\.

With commit 02cab67 Iâve implemented a slight variation on the theme&#8203;: 
Now mro\_get\_linear\_isa returns a modified array that \*does\* include the
class name\, if what the plugin returned did not have it\.

This means mro&#8203;::get\_linear\_isa will also return the same\.
